Improvements to tag-based mining level API

Shnupbups opened the pull request:
This PR introduces improvements upon the tag-based mining level API introduced in #1629, mostly focused around Shears

- Shears now work with mining level tags.
- Vanilla Shears have a mining level of 2, equivalent to that of Iron tools.
- Introduced `fabric:mineable/shears_fast` and `fabric:mineable/shears_slow` block tags, allowing for utilisation of the mining speeds used by leaves and vines respectively - these tags are nested in the already present `fabric:mineable/shears`, so it is not necessary to manually put a block in both. Blocks not present in either fast nor slow will use the standard speed used by wool.
- Added `FabricTool`, an interface implemented by `ToolItem` and `ShearsItem` and extended by `DynamicAttributeTool` - allows for Shears to utilise `ToolMaterial`s, and provides some potentially useful methods.
- Added `FabricShearsItem`, a class that extends `ShearsItem` - provides a simple way to pass a `ToolMaterial` into shears.
- Potentially fixed issues regarding `DynamicAttributeTool`s mining blocks with dynamic mining levels
- Added stuff to the Test Mod which utilises new functionality introduced by this PR

Shnupbups committed:
shears stuff

Juuxel commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricMineableTags.java:
```diff
@@ -43,6 +43,16 @@
+	public static final Tag.Identified<Block> SHEARS_MINEABLE_FAST = register("mineable/shears_fast");
+
+	/**
+	 * Blocks in this tag ({@code #fabric:mineable/shears}) can be effectively mined with shears at a slower than normal rate.
```
```suggestion
	 * Blocks in this tag ({@code #fabric:mineable/shears_slow}) can be effectively mined with shears at a slower than normal rate.
```

Juuxel commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricMineableTags.java:
```diff
@@ -43,6 +43,16 @@
 	public static final Tag.Identified<Block> SHEARS_MINEABLE = register("mineable/shears");
 
+	/**
+	 * Blocks in this tag ({@code #fabric:mineable/shears}) can be effectively mined with shears at a faster than normal rate.
```
```suggestion
	 * Blocks in this tag ({@code #fabric:mineable/shears_fast}) can be effectively mined with shears at a faster than normal rate.
```

Juuxel commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java:
```diff
@@ -0,0 +1,54 @@
+import net.minecraft.item.ToolMaterial;
+import net.minecraft.tag.Tag;
+
+public interface FabricTool {
```
This interface and its methods should be documented.

Shnupbups replied:
Done, let me know if it's sufficient

Juuxel commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java:
```diff
@@ -0,0 +1,54 @@
+		}
+	}
+
+	default boolean isSuitableFor(int miningLevel, BlockState state) {
```
Is this called anywhere in the mining level API? Any tool attributes stuff should stay in that module - all features here should work without the tool attribute api being present.

Shnupbups replied:
While it is only called by the tool attributes module, it is necessary for that module's `DynamicAttributeTool`s to work with dynamic mining levels introduced by this module.

I don't see your point exactly - it only being called by another module in no way means that this module relies on the other. Mining Level will work just fine without Tool Attributes.

Juuxel replied:
My point is that the interface has features that don't work when the module is used alone with just its deps. I guess you could work around that by saying something like "Called in other Fabric API modules" in the JD as an explanation for it, but that feels a bit hacky to me.

Shnupbups replied:
> the interface has features that don't work when the module is used alone with just its deps.

What? The feature of the interface works just fine, it still returns exactly as it should.
This method doesn't require tool attributes at all, it's the other way around. Tool attributes requires this method. This method works just fine without tool attributes.

Juuxel replied:
You'd expect this method to at least do something when overridden, when it does precisely nothing since it's not called. I don't really see why the usual item method doesn't suffice here since you also know the mining level from the same interface.

Shnupbups replied:
That's... actually a good point, don't know why I didn't consider that.

I've updated it so now this method is actually used by the regular `isSuitableFor`, and thus will work when overridden.

Shnupbups committed:
Apply suggestions from code review

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>

Shnupbups committed:
Document FabricTool

Shnupbups committed:
Merge branch '1.17-shears' of https://github.com/Shnupbups/fabric into 1.17-shears

Juuxel commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java:
```diff
@@ -0,0 +1,82 @@
+	 * Returns whether a tool of this type, if it had the mining level provided, would be suitable for mining the provided {@code BlockState}.
+	 * Can be useful in circumstances where {@link net.minecraft.item.Item#isSuitableFor(BlockState)} is not.
+	 * @param miningLevel the mining level of the tool
+	 * @param state the blockstate being mined
```
```suggestion
	 * @param state the block state being mined
```
Small typo fix

Shnupbups committed:
Update fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>

Technici4n commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java:
```diff
@@ -0,0 +1,82 @@
+	 * @return The tool material
+	 */
+	@Nullable
+	default ToolMaterial getToolMaterial() {
```
What about dynamic tools?

Shnupbups replied:
Dynamic Tools return null, as per the javadoc.

Technici4n commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java:
```diff
@@ -0,0 +1,82 @@
+	 * @return The effective blocks
+	 */
+	@Nullable
+	default Tag<Block> getEffectiveBlocks() {
```
What about tools that can work on multiple kinds of blocks, e.g. paxels?

Shnupbups replied:
These can either return a tag that nests `minecraft:mineable/pickaxe`, `minecraft:mineable/axe` and `minecraft:mineable/shovel`, or return null as per the javadoc.

Technici4n replied:
I'm not sure I see the point of these functions then.

Shnupbups replied:
`getToolMaterial` exists so that things like shears can be tied to a tool material, and thus inherit their mining level from it
`getEffectiveBlocks` exists so `isSuitableFor` and `getMiningSpeedMultiplier` can use them
What is so hard to understand about that?

`FabricTool` isn't intended for making fancy multitools or anything, but simply allowing for the creation of simple vanilla-style tools that aren't pickaxes, axes, shovels, or hoes. This API specifically uses it for swords and shears, but it can potentially be used for other things too.

`DynamicAttributeTool` extends `FabricTool`, and it is then ideal for use for making more fancy tools like paxels and drills and whatnot.

Technici4n replied:
Imo what these methods are being used for is not super clear. Is it just metadata? Or does it have some special behavior? What is the effect of overriding them? Perhaps they could be extended a tiny little bit to replace the dynamic tool module entirely?

Shnupbups committed:
whoops haha

forgot to have material mining speed multipliers factor into shears, they do now!

Shnupbups committed:
Merge branch '1.17-shears' of https://github.com/Shnupbups/fabric into 1.17-shears

Shnupbups committed:
whoops

Shnupbups committed:
there we go

shears material speeds should be handled properly now

Shnupbups committed:
fixed test mod shears durability, added textures

Shnupbups committed:
your mum's a checkstyle rule violation

NinjaPhenix commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java:
```diff
@@ -0,0 +1,82 @@
+		ToolMaterial toolMaterial = getToolMaterial();
+
+		if (toolMaterial != null) {
+			return getToolMaterial().getMiningLevel();
```
Why not just use the local variable `toolMaterial` here...

Shnupbups replied:
lol must have missed that when refactoring some things, fixed

NinjaPhenix reviewed (approved):
Looks good imo

Shnupbups committed:
improvements

Shnupbups committed:
Fix #1805 i think

Shnupbups committed:
Merge branch '1.17' into 1.17-shears

Juuxel commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/handlers/TaggedToolsModdedBlocksToolHandler.java:
```diff
@@ -41,13 +42,13 @@
 				int miningLevel = 0; // minimum mining level: the tool is tagged but nothing else is said about it
 				int requiredMiningLevel = entry.getMiningLevel(tag);
-				return requiredMiningLevel >= miningLevel ? ActionResult.SUCCESS : ActionResult.PASS;
+				return miningLevel >= requiredMiningLevel ? ActionResult.SUCCESS : ActionResult.PASS;
```
My fix for this is a bit more comprehensive (and you need to fix the merge conflicts anyway).

Also, are all `ToolItem`s `FabricTool`s? If not, the if check above would have a different meaning

Juuxel replied:
Did I comment just as you fixed the conflicts? :eyes:

Shnupbups replied:
Haha, yeah I fixed the conflicts pretty quickly.

And yes, `ToolItem` implements `FabricTool` courtesy of `ToolItemMixin`. All `ToolItem`s, `ShearsItem`s, and `DynamicAttributeTool`s are `FabricTool`s.

Juuxel commented on fabric-mining-level-api-v1/src/main/resources/data/fabric/tags/blocks/mineable/shears.json:
```diff
@@ -0,0 +1,7 @@
+  "replace": false,
+  "values": [
+    "#fabric:mineable/shears_fast",
+    "#fabric:mineable/shears_slow"
```
Does this work when the fast/slow tags are not defined? (In a plain environment without the testmod)

Shnupbups replied:
Good shout, fixed

Shnupbups committed:
whoopsie

Juuxel commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	/**
+	 * Returns whether a tool of this type, if it had the mining level provided, would be suitable for mining the provided {@code BlockState}.
```
```suggestion
	 * Returns whether a tool of this type, if it had the mining level provided, would be suitable for mining the provided {@link BlockState}.
	 * Automatically called for subclasses of {@link net.minecraft.item.ToolItem}, {@link net.minecraft.item.SwordItem} and {@link net.minecraft.item.ShearsItem}.
```

Juuxel commented on fabric-mining-level-api-v1/src/testmod/resources/assets/fabric-mining-level-api-v1-testmod/lang/en_us.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "item.minecraft.shears": "Iron Shears",
```
Not really needed in a testmod

Shnupbups replied:
eh, neither are things like textures, but i figured 'why not'. after all, it doesn't hurt

Juuxel commented on fabric-mining-level-api-v1/src/testmod/java/net/fabricmc/fabric/test/mininglevel/MiningLevelTest.java:
```diff
@@ -95,13 +140,28 @@ private static void test() {
+
+		checkMiningSpeed(Blocks.VINE, WOODEN_SHEARS, STONE_SHEARS, Items.SHEARS, DIAMOND_SHEARS, NETHERITE_SHEARS, GOLDEN_SHEARS, Items.STICK);
+		checkMiningSpeed(Blocks.WHITE_WOOL, WOODEN_SHEARS, STONE_SHEARS, Items.SHEARS, DIAMOND_SHEARS, NETHERITE_SHEARS, GOLDEN_SHEARS, Items.STICK);
+		checkMiningSpeed(Blocks.OAK_LEAVES, WOODEN_SHEARS, STONE_SHEARS, Items.SHEARS, DIAMOND_SHEARS, NETHERITE_SHEARS, GOLDEN_SHEARS, Items.STICK);
```
I don't really see why these should be printed out. Instead, it might be useful to compare the mining speeds like the tool attribute test.

Shnupbups replied:
Perhaps, I just wanted something quick and dirty to do the trick for debugging purposes and figured that plenty of other test mods spam the log with prints, so it wouldn't be too terrible.

Juuxel commented on fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java:
```diff
@@ -0,0 +1,82 @@
+	 * @param state the block state being mined
+	 * @return Whether the tool is suitable
+	 */
+	default boolean isSuitableFor(int miningLevel, BlockState state) {
```
Also, why is this needed? You know the mining level from this same interface, so you don't really have to do the "if it had the mining level" guesswork.

Shnupbups replied:
See my response to the `ModdedToolsVanillaBlocksToolHandler` comment you made below.

Juuxel commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/handlers/ModdedToolsVanillaBlocksToolHandler.java:
```diff
@@ -55,18 +56,18 @@ private ToolItem getVanillaItem(int miningLevel) {
-			ToolItem vanillaItem = getVanillaItem(miningLevel);
-			return vanillaItem.isSuitableFor(state) ? ActionResult.SUCCESS : ActionResult.PASS;
+			FabricTool vanillaItem = (FabricTool) getVanillaItem(miningLevel);
+			return vanillaItem.isSuitableFor(miningLevel, state) ? ActionResult.SUCCESS : ActionResult.PASS;
```
This doesn't really need to be changed to use the interface's `isSuitableFor` since they return the same value because of the mixin, but this is fine too I guess.

Shnupbups replied:
`getVanillaItem` can only return... well, vanilla items (or rather, items it specifically knows about, which happen to be just vanilla items). That is, for example, wooden, stone, iron, diamond, and netherite pickaxes, of mining levels 0, 1, 2, 3, and 4 respectively.
If `getVanillaItem(5)` is called, it will return the highest one it knows about: netherite, which has a mining level of _4_ and not the desired 5.
Thus, if the original `isSuitableFor(state)` were to be called on the item the method returns, in this case it would be checking if a tool with a mining level of _4_ was suitable for the state... which is not what we want.
So, instead, we use `isSuitableFor(miningLevel, state)` to ensure that we are indeed checking for a mining level of 5.

This then justifies the existence of said method.

Shnupbups committed:
Update fabric-mining-level-api-v1/src/main/java/net/fabricmc/fabric/api/mininglevel/v1/FabricTool.java

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>

Technici4n commented:
I don't think this will ever get merged in its current state. Maybe some bits could be PR'ed separately.
